---
title: "edgeR_normalization"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
#Load the edgeR and statmod libraries
library("edgeR")
library("statmod")
#import data matrix 

# tidyverse core packages
library(tibble)
library(dplyr)
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)
library(purrr)
library(tidyverse)
library(ComplexHeatmap)
library(coseq)
library(Biobase)
library(corrplot)
# tidyverse-friendly packages
library(plotly)
library(ggrepel)
library(GGally)
library(ggvenn)
library(DESeq2)
library("sva")
library("gridExtra")
library(ComplexHeatmap)
library(circlize)
```


```{r}
cts <- as.matrix(read.csv("../readcount.txt",sep="\t",row.names="gene_id"))

#check if the data has correctly been upload

head(cts)

#making a study design file

# set `mock` as reference level

###First samples 

cts_2 <- as.matrix(read.csv("../readcount_RNAseq1.txt",sep="\t",row.names="gene_id"))


head(cts_2)

colnames(cts_2) <- c("R1_4", "R1_5",	"R1_6", "M4",	"M5",	"M6",	"R25_4", "R25_5", "R25_6")

cts_merged <- merge(cts, cts_2, by="row.names")
head(cts_merged)

```

###Remove two mock samples from the first run


```{r}
cts_merged_mkrm <- cts_merged[, -which(names(cts_merged) %in% c("R1_3", "R1_5", "R25_3", "M4","M5"))]


```

###create a study design for ```edgeR```

```{r}

trt <- c(rep("Race1", 2), rep("mock", 3), rep("Race25", 2), rep("Race1", 2), rep("mock", 1), rep("Race25", 3))

#trt <- relevel(trt, ref="mock")

batch <- as.factor(c(rep("batch2", 7), rep("batch1", 6)))

std_design <-data.frame(trt, batch)
std_design$trt <- relevel(std_design$trt, ref = "mock")
row.names(std_design) <- colnames(cts_merged_mkrm)[2:14]


row.names(cts_merged_mkrm) <- cts_merged_mkrm$Row.names
cts_merged_mkrm$Row.names <- NULL

group <- factor(paste(std_design$trt,std_design$batch,sep="."))
std_design <- cbind(std_design,Group=group)
```

###create ```edgeR``` object and label samples using ```std_design```

```{r}
combine_list <- DGEList(counts=cts_merged_mkrm,group=group)
colnames(combine_list) <- row.names(std_design)

```




#Retain genes only if it is expressed at a minimum level

```{r}
keep <- rowMeans(cts_merged_mkrm) > 20
combine_list <- combine_list[keep, , keep.lib.sizes=FALSE]
#View a summary of the normalized counts
summary(keep)

```

#Plot the library sizes before normalization
```{r}
barplot(combine_list$samples$lib.size*1e-6, names=1:ncol(combine_list), ylab="Library size (millions)")
```


#Use TMM or upper quantile normalization to eliminate composition biases

```{r}
combine_list_TMM <- calcNormFactors(combine_list, method = "TMM")
combine_TMM_norm.factors <- combine_list_TMM$samples


combine_list_UQ <- calcNormFactors(combine_list, method = "upperquartile")
combine_UQ_norm.factors <- combine_list_UQ$samples
```

###MDS plots with two different normalization. Both seems very similar. 
```{r}
plotMDS(combine_list_TMM, col=c(rep(1, 2), rep(2, 3), rep(3, 2), rep(4, 3), rep(5, 1), rep(6, 2)))
plotMDS(combine_list_UQ, col=c(rep(1, 2), rep(2, 3), rep(3, 2), rep(4, 3), rep(5, 1), rep(6, 2)))
```



##Batch Correction
=========================================================================================
```
Introduction to Batch correction
https://rnabio.org/module-03-expression/0003/05/01/Batch-Correction/

We highly recommend reading the entire ComBat-Seq manuscript by Yuqing Zhang, Giovanni Parmigiani, and W Evan Johnson. This manuscript does a beautiful job of briefly introducing the concept of batch correction and the differences between normalization and batch correction. If you find this exercise helpful in your research, please cite the ComBat-Seq paper (PMID: 33015620).

In particular, this excerpt covers the basics:

Genomic data are often produced in batches due to logistical or practical restrictions, but technical variation and differences across batches, often called batch effects, can cause significant heterogeneity across batches of data. Batch effects often result in discrepancies in the statistical distributions across data from different technical processing batches, and can have unfavorable impact on downstream biological analysis …

Batch effects often cannot be fully addressed by normalization methods and procedures. The differences in the overall expression distribution of each sample across batch may be corrected by normalization methods, such as transforming the raw counts to (logarithms of) CPM, TPM or RPKM/FPKM, the trimmed mean of M values (TMM), or relative log expression (RLE). However, batch effects in composition, i.e. the level of expression of genes scaled by the total expression (coverage) in each sample, cannot be fully corrected with normalization. … while the overall distribution of samples may be normalized to the same level across batches, individual genes may still be affected by batch-level bias.



```
####preparing study design for Combat analysis

```{r}
replicates = c(1, 2, 1, 2, 3, 1, 2, 4, 5, 4, 4, 5, 6)
std_design_sva <-data.frame(trt, batch, replicates)
```

####Combat analysis

```{r}

groups = sapply(as.character(trt), switch, "Race1" = 1, "mock" = 3, "Race25"= 2, USE.NAMES = F)
groups
corrected_data = ComBat_seq(counts = as.matrix(cts_merged_mkrm), batch = batch, group = groups)

write.csv(corrected_data, file="20220304_corrected_data.csv")
```



###create ```edgeR``` object and label samples using ```std_design```

```{r}
combine_list_corrected <- DGEList(counts=corrected_data, group=group)
group
colnames(combine_list_corrected) <- row.names(std_design)

```




#Retain genes only if it is expressed at a minimum level

```{r}
#keep <- filterByExpr(combine_list_corrected)
combine_list_corrected <- combine_list_corrected[keep, , keep.lib.sizes=FALSE]
#View a summary of the normalized counts
summary(keep)

```

#Plot the library sizes before normalization
```{r}
barplot(combine_list_corrected$samples$lib.size*1e-6, names=1:ncol(combine_list_corrected), ylab="Library size after batch corrected (millions)")
```


#Use TMM or upper quantile normalization to eliminate composition biases

```{r}
combine_list_correctedTMM <- calcNormFactors(combine_list_corrected, method = "TMM")
combine_correctedTMM_norm.factors <- combine_list_correctedTMM$samples


combine_list_correctedUQ <- calcNormFactors(combine_list_corrected, method = "upperquartile")
combine_list_correctedUQ_norm.factors <- combine_list_correctedUQ$samples
```

###MDS plots with two different normalization. Both seems very similar. 
```{r}
plotMDS(combine_list_correctedTMM, col=std_design_sva$replicates)
std_design_sva$replicates
plotMDS(combine_list_correctedUQ, col=c(rep(1, 2), rep(2, 3), rep(3, 2), rep(4, 2), rep(5, 1), rep(6, 3)))
```

##using batch correction using Limma

```{r}
CPM_TMM <- cpm(combine_list_TMM, normalized.lib.sizes=TRUE)
CPM_TMM_br <- removeBatchEffect(CPM_TMM, batch=std_design$batch)


plotMDS(CPM_TMM_br, col=c(rep(1, 2), rep(2, 3), rep(3, 2), rep(4, 2), rep(5, 1), rep(6, 3)))
```



```{r}
trt 
batch

design_logfc <- model.matrix(~batch+batch:trt)

logFC <- predFC(combine_list_correctedTMM, design_logfc, prior.count=1, dispersion=0.05)

logFC[3,] 

cor(logFC)

```

###GLM estimates of dispersion

```{r}
design_mat <- model.matrix(~std_design$batch + std_design$trt)


rownames(design_mat) <- rownames(std_design)



design_mat
combine_list_correctedTMM <- calcNormFactors(combine_list_correctedTMM)

combine_list_correctedTMM <- estimateDisp(combine_list_correctedTMM, design_mat, robust=TRUE)

#common dispersion
combine_list_correctedTMM$common.dispersion

plotBCV(combine_list_correctedTMM)

correctedTMM_fit <- glmQLFit(combine_list_correctedTMM, design_mat, robust=TRUE)
plotQLDisp(correctedTMM_fit)

```



```{r}

# compare (group 1 - group 2) to 0:
# this is equivalent to comparing group 1 to group 2
lrt_batch1vs2 <- glmLRT(correctedTMM_fit, contrast=c(1, -1,0, 0))
lrt_mock_R1 <- glmLRT(correctedTMM_fit, coef = 3)
lrt_mock_R25 <- glmLRT(correctedTMM_fit, coef = 4)
lrt_R1_R25 <- glmLRT(correctedTMM_fit, contrast=c(0, 0,-1, 1))

#top genes between mock and R1
topTags(lrt_mock_R1, n=10)

#top genes between mock and R25
topTags(lrt_mock_R25, n=10)

#top genes between R1 and R25

topTags(lrt_R1_R25, n=10)

hist(lrt_batch1vs2$table[,"PValue"], breaks=50)

hist(lrt_mock_R1$table[,"PValue"], breaks=50)


hist(lrt_mock_R25$table[,"PValue"], breaks=50)

hist(lrt_R1_R25$table[,"PValue"], breaks=50)
```




```{r}
getDE_genes <- function(edgR_DElist, padj_cutoff=0.05, comparison){
  
  DE_dataframe <- as.data.frame(topTags(edgR_DElist, n=nrow(edgR_DElist$table))$table)
  DE_dataframe <- DE_dataframe[DE_dataframe$FDR <= padj_cutoff, ]
  colnames(DE_dataframe) <- c(paste(colnames(DE_dataframe), "_", comparison, sep = ""))
  DE_dataframe$geneId <- row.names(DE_dataframe)
return(DE_dataframe)
}

```


```{r}

DE_mock_R1 <- getDE_genes(lrt_mock_R1, padj_cutoff=0.05, "mock_R1")

DE_mock_R25 <- getDE_genes(lrt_mock_R25, padj_cutoff=0.05, "mock_R25")

DE_mock_R1_25 <- getDE_genes(lrt_R1_R25, padj_cutoff=0.05, "mock_R1_25")

DE_batch1vs2 <- getDE_genes(lrt_batch1vs2, padj_cutoff=0.05, "mock_R1_25")

```


```{r}
EdgeR_list <- list(
  lrt_mock_R1_names = row.names(DE_mock_R1), 
  lrt_mock_R25_names = row.names(DE_mock_R25), 
  Race1_Race25_names = row.names(DE_mock_R1_25),
  batch1_2_names = row.names(DE_batch1vs2)
  )


ggvenn(
  EdgeR_list, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4
)

```



####use corrected data for DEseq 
```{r}
dse <- DESeqDataSetFromMatrix(corrected_data,
                              colData = std_design,
                              design =~batch + trt)
```

```{r}
# use keep from EdgeR 

dse <- dse[keep,]

```


#### vsd tranformation to check the samples 

```{r}
vsd <- vst(dse,blind=TRUE)
```

####check data using a PCA

####make a PCA data for ggplot
```{r}
plot_DEseq <- plotPCA(vsd,intgroup=c("trt"),returnData=TRUE)
DEseq_PCA <- ggplot(plot_DEseq, aes(x = PC1,y=PC2, col=trt,label=name)) + 
  geom_point() + geom_text(alpha=0.6)

```


```{r}

#edgeR_MDS <- plotMDS(combine_list_correctedTMM, col=c(rep(1, 2), rep(2, 3), rep(3, 2), rep(4, 2), rep(5, 1), rep(6, 3)))
DEseq_PCA

plotMDS(combine_list_correctedTMM, col=c(rep(1, 2), rep(2, 3), rep(3, 2), rep(4, 2), rep(5, 1), rep(6, 3)))

```



####differential gene expression analysis

```{r}
dds <- DESeq(dse)
plotDispEsts(dds)
resultsNames(dds)
```


#### generate results table for different combinations

```{r}

DEseq_Race1_vs_mock <- results(dds, name="trt_Race1_vs_mock")

DEseq_Race25_vs_mock <- results(dds, name="trt_Race25_vs_mock")

DEseq_Race1_Race25 <- results(dds, c('trt', "Race1", "Race25"))

DEseq_batch2_batch2<- results(dds, name = "batch_batch2_vs_batch1" )

```


####create a function to filter signficant genes

```{r}
getDE_genes <- function(res_counts, padj_cutoff=0.05, comparison){
  DE_dataframe <- as.data.frame(res_counts[!is.na(res_counts$padj) & res_counts$padj <= padj_cutoff, ])
  DE_dataframe <- DE_dataframe[, c(1, 2, 6)]
  colnames(DE_dataframe) <- c(paste(colnames(DE_dataframe), "_", comparison, sep = ""))
  DE_dataframe$geneId <- row.names(DE_dataframe)
  return(DE_dataframe)
}
```
####filter signficant genes

```{r}
DEseq_Race1_vs_mock_sig <- getDE_genes(DEseq_Race1_vs_mock, padj_cutoff=0.05, "Race1_vs_mock")
nrow(DEseq_Race1_vs_mock_sig)
DEseq_Race25_vs_mock_sig <- getDE_genes(DEseq_Race25_vs_mock, padj_cutoff=0.05, "Race25_vs_mock")
nrow(DEseq_Race25_vs_mock_sig)
DEseq_Race1_Race25_sig <- getDE_genes(DEseq_Race1_Race25, padj_cutoff=0.05, "Race1_Race25")
nrow(DEseq_Race1_Race25_sig)
DEseq_batch2_batch2_sig <- getDE_genes(DEseq_batch2_batch2, padj_cutoff=0.05, "batch2_batch2")
nrow(DEseq_batch2_batch2_sig)
```

###Note on 3/30/2022: There was an error in the code. Therefore, DE list was not correct. I have corrected the code. Checked exported individial DE list for each comparison and counted numbers and compared it with combined Venn diagram. This seems correct.

```
#DEseq_Race1_vs_mock 
write.csv(as.data.frame(DEseq_Race1_vs_mock), "20220328_DEseq_Race1_vs_mock.csv")
#DEseq_Race25_vs_mock 
write.csv(as.data.frame(DEseq_Race25_vs_mock), "20220328_DEseq_Race25_vs_mock.csv")
#DEseq_Race1_Race25
write.csv(as.data.frame(DEseq_Race1_Race25), "20220328_DEseq_Race1_Race25.csv")
#DEseq_batch2_batch2
write.csv(as.data.frame(DEseq_batch2_batch2), "20220328_DEseq_batch2_batch2.csv")
```


####check the overlap of the different genes 

```{r}
DEseq_list <- list(
  Race1_vs_mock = row.names(DEseq_Race1_vs_mock_sig), 
  Race25_vs_mock = row.names(DEseq_Race25_vs_mock_sig), 
  Race1_Race25 = row.names(DEseq_Race1_Race25_sig),
  Batch1and2 =row.names(DEseq_batch2_batch2_sig)
)


ggvenn(
  DEseq_list, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4
)


nrow(DEseq_batch2_batch2_sig)
```

#####after corrections
```{r}
d_corrected <- plotCounts(dds, gene="Glyma.04G121700", intgroup="trt", normalized = TRUE,
                returnData=TRUE)
d_corrected

ggplot(d_corrected, aes(x=trt, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))

```



####house keeping gene expression

####create a function to return a plot with gene expression for more than one gene to modify DEseq plotCount funciton



```{r}
gene_expression_graph <- function(res_counts, gene_names){
  house_keeping_names = read.table(textConnection(gene_names),header=FALSE)
  datalist = list()
  
  for (i in 1:dim(house_keeping_names)[1]) {
    gene_name <- house_keeping_names$V1[i]
    print(as.character(gene_name))
    dat <- plotCounts(res_counts, gene=as.character(gene_name), intgroup="trt", normalized = TRUE,
                      returnData=TRUE)
    dat$gene_name <- gene_name
    datalist[[i]] <- dat
    
  }
  
  gene_expression <- do.call(rbind, datalist)
  house_keeping_gg <- ggplot(gene_expression, aes(x=trt, y=count)) + 
    geom_point(position=position_jitter(w=0.1,h=0)) + 
    scale_y_log10(breaks=c(25,100,400))
  
  house_keeping_gg <- house_keeping_gg + facet_wrap( ~ gene_name, ncol=2)
  
  return(house_keeping_gg)
  
  
}

```


```{r}
house_keeping <- c("
Glyma.20G141600
Glyma.12G020500
Glyma.12G051100
Glyma.20G136000
Glyma.12G024700
Glyma.08G182200
                ")


#gene_expression_graph(dds, house_keeping)

```



####pathogen induced genes

```{r}
pathogen_genes <- c("
Glyma.02G028100
Glyma.02G028600
Glyma.02G268200
Glyma.04G131100
Glyma.18G177000
Glyma.19G011700
Glyma.19G188100
Glyma.16G195600
")

#gene_expression_graph(dds, pathogen_genes )

```



####gene used in DAPseq

```{r}
DAPseq_genes <- c("
Glyma.06G125600
Glyma.10G051500
Glyma.10G142200
Glyma.06G320700
                ")


#gene_expression_graph(dds, DAPseq_genes)

```






####merge significantly changed genes

```{r}
all_sign_genes <- Reduce(function(x, y) merge(x, y, all=TRUE, by ="geneId"), 
       list(DEseq_Race1_vs_mock_sig , 
            DEseq_Race25_vs_mock_sig, 
            DEseq_Race1_Race25_sig,
            DEseq_batch2_batch2_sig))

head(all_sign_genes)
```

###Normalized counts 


```{r}
DEseq_norm <- data.frame(counts(dds, normalized=TRUE))
DEseq_norm <- DEseq_norm + 1
head(DEseq_norm)

```

###to remove genes affact by the batch alone 
```{r}
DEseq_Race1_vs_mock_sig_br <- DEseq_Race1_vs_mock_sig[!row.names(DEseq_Race1_vs_mock_sig) %in% row.names(DEseq_batch2_batch2_sig), ]
nrow(DEseq_Race1_vs_mock_sig)
nrow(DEseq_Race1_vs_mock_sig_br)

DEseq_Race25_vs_mock_sig_br <- DEseq_Race25_vs_mock_sig[!row.names(DEseq_Race25_vs_mock_sig) %in% row.names(DEseq_batch2_batch2_sig),]
DEseq_Race1_Race25_sig_br <- DEseq_Race1_Race25_sig[!row.names(DEseq_Race1_Race25_sig) %in% row.names(DEseq_batch2_batch2_sig),]
```



###get log fold change and significance
```{r}
DEseq_norm$geneId <- rownames(DEseq_norm)

DEseq_norm <- DEseq_norm[, c("geneId", "M1", "M2", "M3", "M6", "R1_1", "R1_2", "R1_4", "R1_6", "R25_1", "R25_2", "R25_4", "R25_5", "R25_6")]

DEseq_norm <- DEseq_norm %>%
  
  mutate(DE_R1_mock = if_else(rownames(DEseq_norm) %in% row.names(DEseq_Race1_vs_mock_sig_br), "yes", "no")) %>%
  mutate(DE_R25_mock = if_else(rownames(DEseq_norm) %in% row.names(DEseq_Race25_vs_mock_sig_br), "yes", "no")) %>%
  mutate(DE_R1_R25 = if_else(rownames(DEseq_norm) %in% row.names(DEseq_Race1_Race25_sig_br), "yes", "no")) %>%
  rowwise()%>%
  mutate(meanM = sum(M1, M2, M3, M6)/4, meanR1 = sum(R1_1, R1_2, R1_4, R1_6)/4,  meanR25 = sum(R25_1, R25_2, R25_4, R25_5, R25_6)/5) %>%
  mutate(R1_LFC = log2(meanR1/meanM), R25_LFC = log2(meanR25/meanM))


head(as.data.frame(DEseq_norm))
```



###pathogen genes

```{r}
pathogen_names = read.table(textConnection(pathogen_genes), header=FALSE)
as.data.frame(DEseq_norm[DEseq_norm$geneId %in% pathogen_names$V1, ])
```

###import captureseq data

```{r}
capture_seq <- read.csv("Capture_seq.csv")


head(capture_seq)


```

###merging RNAseq and captureseq entire dataset 

```{r}
RNAseq_capture_all <- merge(capture_seq, DEseq_norm, by = "geneId")

head(RNAseq_capture_all)

```

####renameing and selecting the columns 

```{r}
RNAseq_capture_all <- RNAseq_capture_all %>% 
  replace_na(list(log2FoldChange_Race1_vs_mock = 0, log2FoldChange_Race25_vs_mock = 0)) %>%
  rename(LFC_Capture_Race1_mock = log2FoldChange_Race1_vs_mock,
         LFC_Capture_Race25_mock = log2FoldChange_Race25_vs_mock,
         LFC_RNAseq_Race1_mock = R1_LFC,
         LFC_RNAseq_Race25_mock = R25_LFC)

write.csv(RNAseq_capture_all, file="20220401_RNAseq_capture_all.csv")

RNAseq_capture_matrix <- RNAseq_capture_all %>% 
  select(-geneId) %>% 
  select(c( "LFC_Capture_Race1_mock", "LFC_Capture_Race25_mock", "LFC_RNAseq_Race1_mock", "LFC_RNAseq_Race25_mock")) %>%
  
  as.matrix()

rownames(RNAseq_capture_matrix) <- RNAseq_capture_all$geneId
#colnames(RNAseq_capture_matrix) <- c( "LFC_Capture_Race1_mock", "LFC_Capture_Race25_mock", "LFC_RNAseq_Race1_mock")


#"DE_R1_mock", "DE_R25_mock", "Race25_vs_mock_sig", "Race1_vs_mock_sig", "LFC_Capture_Race1_mock"
```


```{r}

Heatmap(RNAseq_capture_matrix, km = 5, col = colorRamp2(c(-4, 0, 4), c("green", "white", "red")), show_column_names = TRUE, show_row_names = FALSE ) +
  Heatmap(as.matrix(RNAseq_capture_all[, c(16, 17, 31, 32)]),  name = "Statical significance", show_row_names = FALSE, width = unit(20, "mm")) 

```




###compare RNAseq and Captureseq

```{r}
#Race1 and Mock 
table(RNAseq_capture_all[, c(17, 31)])
#Race25 and Mock 
table(RNAseq_capture_all[, c(16, 32)])


```

```{r}
hub_genes <- c("Glyma.01G044600
                  Glyma.01G045500
                   Glyma.01G047900
                   Glyma.01G049100
                   Glyma.01G088200
                   Glyma.01G128100
                   Glyma.02G019100
                   Glyma.02G071500
                   Glyma.02G162600
                   Glyma.02G232600
                   Glyma.02G258300
                   Glyma.03G042700
                   Glyma.03G163100
                   Glyma.03G225200
                   Glyma.04G033300
                   Glyma.04G170600
                   Glyma.04G238300
                   Glyma.04G249000
                   Glyma.05G018100
                   Glyma.05G144500
                   Glyma.06G040900
                   Glyma.06G085700
                   Glyma.06G125600
                   Glyma.06G149900
                   Glyma.07G052100
                   Glyma.07G126800
                   Glyma.07G198900
                   Glyma.07G235900
                   Glyma.07G262700
                   Glyma.08G056800
                   Glyma.08G135200
                   Glyma.08G202300
                   Glyma.09G014100
                   Glyma.09G109700
                   Glyma.10G118900
                   Glyma.10G142200
                   Glyma.11G035900
                   Glyma.11G043700
                   Glyma.11G049300
                   Glyma.11G096600
                   Glyma.11G242200
                   Glyma.12G022700
                   Glyma.12G137700
                   Glyma.12G216800
                   Glyma.13G062900
                   Glyma.13G117600
                   Glyma.13G139000

")
hubgenes_names = read.table(textConnection(hub_genes),header=FALSE)
hubgenes_names_df <- RNAseq_capture_all[RNAseq_capture_all$geneId %in% hubgenes_names$V1, ]

write.csv(hubgenes_names_df, file="20220401_hubgenes_names_df_Deseq.csv", row.names = FALSE)



hubgenes_matrix <- hubgenes_names_df %>% 
  select(c( "LFC_Capture_Race1_mock", "LFC_Capture_Race25_mock", "LFC_RNAseq_Race1_mock", "LFC_RNAseq_Race25_mock")) %>%
  as.matrix()

rownames(hubgenes_matrix) <- hubgenes_names_df$geneId




Heatmap(hubgenes_matrix, km = 5, col = colorRamp2(c(-4, 0, 4), c("green", "white", "red")), show_column_names = TRUE, show_row_names = FALSE ) +
  Heatmap(as.matrix(hubgenes_names_df[, c("DE_R1_mock", "Race1_vs_mock_sig", "DE_R25_mock","Race25_vs_mock_sig")]),  name = "Statical significance", show_row_names = FALSE, width = unit(20, "mm")) 




```


```{r}
x_hub <- list(
  RNAseq_Race1_mock = row.names(hubgenes_names_df[hubgenes_names_df$DE_R1_mock == "yes",]), 
  RNAseq_Race25_mock = row.names(hubgenes_names_df[hubgenes_names_df$DE_R25_mock == "yes", ]), 
  Captueseq_Race1_mock = row.names(hubgenes_names_df[hubgenes_names_df$Race1_vs_mock_sig == "yes",]), 
  Captueseq_Race25_mock = row.names(hubgenes_names_df[hubgenes_names_df$Race25_vs_mock_sig == "yes", ])
)


ggvenn(
  x_hub, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4
)


#Race1 and Mock 
table(hubgenes_names_df[, c("DE_R1_mock", "Race1_vs_mock_sig")])
#Race25 and Mock 
table(hubgenes_names_df[, c("DE_R25_mock","Race25_vs_mock_sig")])


```

###compare RNAseq and Captureseq

```{r}

pathogen_names_df <- RNAseq_capture_all[RNAseq_capture_all$geneId %in% pathogen_names$V1, ]
pathogen_gene_matrix <- pathogen_names_df %>% 
  select(c( "LFC_Capture_Race1_mock", "LFC_Capture_Race25_mock", "LFC_RNAseq_Race1_mock", "LFC_RNAseq_Race25_mock")) %>%
  as.matrix()

rownames(pathogen_gene_matrix) <- pathogen_names_df$geneId



Heatmap(pathogen_gene_matrix, km = 5, col = colorRamp2(c(-4, 0, 4), c("green", "white", "red")), show_column_names = TRUE, show_row_names = FALSE) +
  Heatmap(as.matrix(pathogen_names_df[, c("DE_R1_mock", "Race1_vs_mock_sig", "DE_R25_mock","Race25_vs_mock_sig")]),  name = "Statical significance", show_row_names = FALSE, width = unit(20, "mm")) 

#Race1 and Mock 
table(pathogen_names_df[, c("DE_R1_mock", "Race1_vs_mock_sig")])
#Race25 and Mock 
table(pathogen_names_df[, c("DE_R25_mock","Race25_vs_mock_sig")])


```

###house keeping genes


```{r}




house_keeping_names = read.table(textConnection(house_keeping), header=FALSE)

#house_keeping

as.data.frame(RNAseq_capture_all[RNAseq_capture_all$geneId %in% house_keeping_names$V1, ])


house_keeping_names_df <- RNAseq_capture_all[RNAseq_capture_all$geneId %in% house_keeping_names$V1, ]
house_keeping_matrix <- house_keeping_names_df %>% 
  select(c( "LFC_Capture_Race1_mock", "LFC_Capture_Race25_mock", "LFC_RNAseq_Race1_mock", "LFC_RNAseq_Race25_mock")) %>%
  as.matrix()

rownames(house_keeping_matrix) <- house_keeping_names_df$geneId



Heatmap(house_keeping_matrix, km = 5, col = colorRamp2(c(-4, 0, 4), c("green", "white", "red")), show_column_names = TRUE, show_row_names = FALSE) +
  Heatmap(as.matrix(house_keeping_names_df[, c("DE_R1_mock", "Race1_vs_mock_sig", "DE_R25_mock","Race25_vs_mock_sig")]),  name = "Statical significance", show_row_names = FALSE, width = unit(20, "mm")) 

#Race1 and Mock 
table(house_keeping_names_df[, c("DE_R1_mock", "Race1_vs_mock_sig")])
#Race25 and Mock 
table(house_keeping_names_df[, c("DE_R25_mock","Race25_vs_mock_sig")])




```


###pathogen genes


```{r}

#pathogen genes

as.data.frame(RNAseq_capture_all[RNAseq_capture_all$geneId %in% pathogen_names$V1, ])

house_keeping_names = read.table(textConnection(house_keeping), header=FALSE)

#house_keeping

as.data.frame(RNAseq_capture_all[RNAseq_capture_all$geneId %in% house_keeping_names$V1, ])
```

##get EdgeR normalized data for clustering

#Use TMM or upper quantile normalization to eliminate composition biases

```{r}
combine_list_correctedTMM <- calcNormFactors(combine_list_corrected, method = "TMM")
combine_correctedTMM_norm.factors <- combine_list_correctedTMM$samples


combine_list_correctedUQ <- calcNormFactors(combine_list_corrected, method = "upperquartile")
combine_list_correctedUQ_norm.factors <- combine_list_correctedUQ$samples
```


```{r}


combine_CTF_normalized_corrected <- sweep(corrected_data, 2, combine_correctedTMM_norm.factors$norm.factors, "/")



combine_CUF_normalized_corrected <- sweep(corrected_data, 2, combine_list_correctedUQ_norm.factors$norm.factors, "/")
```

###get unique gene list for DEseq DE genes 

```{r}
Uniq_DEseq_list <- list(
  Race1_vs_mock = row.names(DEseq_Race1_vs_mock_sig_br), 
  Race25_vs_mock = row.names(DEseq_Race25_vs_mock_sig_br) 
  #Race1_Race25 = row.names(DEseq_Race1_Race25_sig_br)
)
ggvenn(
  Uniq_DEseq_list , 
  fill_color = c("#0073C2FF", "#EFC000FF"),
  stroke_size = 0.5, set_name_size = 4
)


Uniq_DEseq <- unique(unlist(Uniq_DEseq_list, use.names = FALSE))

length(Uniq_DEseq)
```

```{r}


write.csv(DEseq_norm, file="20220401_DEseq_norm_bothbatch_all.csv", row.names = FALSE)

DE_DEseq <- DEseq_norm[DEseq_norm$geneId %in% Uniq_DEseq, ]

write.csv(DE_DEseq, file="20220401_DE_DEseq_norm.csv")


head(DE_DEseq)

```



###get a subset of genes for CTF and CUF normalized counts

```{r}


DEseq_DE_CUF_normalized <- combine_CUF_normalized_corrected[rownames(combine_CUF_normalized_corrected) %in% Uniq_DEseq, ]

DEseq_DE_CTF_normalized <-  combine_CTF_normalized_corrected[rownames(combine_CTF_normalized_corrected) %in% Uniq_DEseq, ]

#write.csv(DEseq_DE_CUF_normalized, file="DEseq_DE_CUF_normalized.csv")
#write.csv(DEseq_DE_CTF_normalized, file="DEseq_DE_CTF_normalized.csv")
```

###Kmean clustering with two different type of normalization 

```

set.seed(1)

DE_CUF_runLogCLR <- coseq(combine_CUF_normalized_corrected, K=2:25, transformation="logclr",norm="none", 
                   meanFilterCutoff=0, model="kmeans",
                   nstart=1, iter.max=15)
set.seed(1)

DE_CTF_runLogCLR <- coseq(DEseq_DE_CTF_normalized, K=2:25, transformation="logclr",norm="none", 
                   meanFilterCutoff=0, model="kmeans",
                   nstart=1, iter.max=15)


summary(DE_CUF_runLogCLR)

summary(DE_CTF_runLogCLR)
```



### plot clustering 

```

##
plot(DE_CUF_runLogCLR, graphs="boxplots")
#plot(DE_CUF_runLogCLR, graphs="boxplots", conds=trt)
#plot(DE_CTF_runLogCLR, graphs="boxplots", conds=trt)
```



### plot clustering using sum of the replicates

```

plot(DE_CUF_runLogCLR, graphs="boxplots", conds=std_design$trt, average_over_conds=TRUE)

plot(DE_CTF_runLogCLR, graphs="boxplots", conds=std_design$trt, average_over_conds=TRUE)
```



